qiskit/circuit/operation.py

```diff
@@ -0,0 +1,62 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Quantum Operation Mixin."""
+
+from abc import ABC
+from qiskit.circuit.exceptions import CircuitError
+
+
+class Operation(ABC):
+    """Quantum Operation Mixin Class.
+    For objects that can be added to a :class:`~qiskit.circuit.QuantumCircuit`.
+    These objects include :class:`~qiskit.circuit.Gate`, :class:`~qiskit.circuit.Reset`,
+    :class:`~qiskit.circuit.Barrier`, :class:`~qiskit.circuit.Measure`,
+    and operators such as :class:`~qiskit.quantum_info.Clifford`.
+    """
+
+    # pylint: disable=unused-argument
+    def __new__(cls, *args, **kwargs):
+        if cls is Operation:
+            raise CircuitError("An Operation mixin should not be instantiated directly.")
+        return super().__new__(cls)
+
+    def __init__(self, name, num_qubits, num_clbits, params):
+        self._name = name
+        self._num_qubits = num_qubits
+        self._num_clbits = num_clbits
+        self._params = params
+
+    @property
```

Comment:
Out of curiosity, why were these removed as `@abstractmethod`s ? Having these as non-abstract method means we're standardizing classes implementing `Operation` as having a `_name`, `_num_qubits`, ... which doesn't seem obviously better than standardizing `name`, `num_qubits`, ....

It also means that answering the question of "Is this a bare `Operation` " from @ewinston 's comment https://github.com/Qiskit/qiskit-terra/pull/7087/files?&file-filters%5B%5D=.py#r739619165 has to be done at class instantiation time (whereas an `@abstractmethod` is checked at class definition time) which, given that we're expecting to make large numbers of these objects, could add up to a substantial runtime cost.

Reply:
@kdk, yes, we did initially have these as `@abstractmethod` s as per your initial code design, but then we saw that we need to reimplement the same getters for (`name`, `num_clbits`, `num_qubits`, etc.) everywhere (e.g., in `Measure`, `Clifford`, `CNOTDihedral`, etc.), and it seemed cleaner to implement them just once as part of `Operation`.

We agree that it would be much better to check that an `Operation` cannot be instantiated directly at compile time vs. at runtime, but we don't know how to do this, with none of the methods being abstract. We would be happy to hear suggestions. 

@ShellyGarion has renamed `_params` to `_operands` as you have requested, though at this point `_operands` are completely unused anywhere in the code. What was the motivation for including `params`/`operands` as part of the `Operation`? Or would you suggest to remove `operands` from `Operation` altogether?

Reply:
Thanks for checking. I believe in this case, having these as class attributes for the cases where they are fixed is sufficient to satisfy them being `@abstractmethod` s. I was wrong though in thinking that the check would happen at class definition time instead of instantiation time, it seems to happen at instantiation time regardless.

```
>>> from abc import ABC, abstractmethod
>>> class base(ABC):
...    @property
...    @abstractmethod
...    def foo(self):
...      pass
...
>>> class B(base):
...   def __init__(self):
...     self.foo = 'bar'
...
>>> B()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't instantiate abstract class B with abstract methods foo
>>> class C(base):
...   foo = 'bar'
...
...
>>> C()
<__main__.C object at 0x10b908e80>
```

My assumption in that in most of these cases,